Add ComparisonValidator to validate comparison of any objects

ChaelCodes opened the pull request:
### Summary

We allow for compare validations in NumericalityValidator, but these
only work on numbers. There are various comparisons people may want
to validate, from dates to strings, to custom comparisons.

```
validates_comparison_of :end_date, greater_than: :start_date
```


### Other Information

My main reason for creating this validator is the number of workarounds for validating that a start_date begins before an end_date.

- https://stackoverflow.com/questions/12198886/how-do-i-validate-start-date-and-end-date
- https://github.com/codegram/date_validator
- https://stackoverflow.com/questions/33639833/validate-that-an-end-date-is-later-than-a-start-date-in-rails-4

Instead of making something specific like a date validator, I thought it would be helpful to create a generic comparison validator, to support a variety of circumstances, including custom comparables.

ChaelCodes committed:
Add ComparisonValidator to Rails to support validations between two comparable values.

We allow for compare validations in NumericalityValidator, but these
only work on numbers. There are various comparisons people may want
to validate, from dates to strings, to custom comparisons.

```
validates_comparison_of :end_date, greater_than: :start_date
```

Refactor NumericalityValidator to share module Comparison with ComparabilityValidator
* Move creating the option_value into a reusable module
* Separate COMPARE_CHECKS which support compare functions and accept values
* Move odd/even checks to NUMBER_CHECKS as they can only be run on numbers

rafaelfranca commented on guides/source/active_record_validations.md:
````diff
@@ -365,6 +365,36 @@ end
+
+```ruby
+class Promotion < ApplicationRecord
+  validates :start_date, compare: { greater_than: :end_date }
````
I think it needs to be `validates :start_date, comparison: { greater_than: :end_date }` given the name of the validator, doesn't it?

ChaelCodes replied:
Yes! Sorry about that. Compare was the old name.

rafaelfranca commented on activemodel/lib/active_model/validations/comparison.rb:
```diff
@@ -0,0 +1,98 @@
+module ActiveModel
+  module Validations
+    class ComparisonValidator < EachValidator # :nodoc:
+      # Options have a check they run against
```
```suggestion
```

rafaelfranca commented on activemodel/lib/active_model/validations/comparison.rb:
```diff
@@ -0,0 +1,98 @@
+      def error_options(value, option_value)
+        error_options = options.except(*CHECKS.keys)
+        error_options[:value] = value
+        error_options.merge!(count: option_value)
```
```suggestion
        error_options[:count] = option_value
```

rafaelfranca commented on activemodel/test/cases/validations/comparison_validation_test.rb:
```diff
@@ -0,0 +1,242 @@
+  end
+
+  private
+    def invalid!(values, error = nil)
```
```suggestion
    def assert_invalid_values(values, error = nil)
```

rafaelfranca commented on activemodel/test/cases/validations/comparison_validation_test.rb:
```diff
@@ -0,0 +1,242 @@
+      end
+    end
+
+    def valid!(values)
```
```suggestion
    def assert_valid_values(values)
```

ChaelCodes replied:
This was intended to match NumericalityValidator's tests:
https://github.com/rails/rails/blob/7101489293ece071013417aeed2da3aa1b0927c9/activemodel/test/cases/validations/numericality_validation_test.rb#L313-L325

It looks like the most common solution in the validation tests is assert_predicate instead of using a helper method.
https://github.com/rails/rails/blob/7101489293ece071013417aeed2da3aa1b0927c9/activemodel/test/cases/validations/format_validation_test.rb#L30-L34

Maybe there's a good argument for refactoring the validation tests to match a common standard? In a separate PR? I know Rails has a [policy against cosmetic changes](https://github.com/rails/rails/pull/13771#issuecomment-32746700) but creating some form of test helper, and refactoring all the validator tests feels larger than a cosmetic change?

rafaelfranca commented on guides/source/active_record_validations.md:
```diff
@@ -365,6 +365,36 @@ end
+### `comparison`
+
+This check will validate a comparison between any two comparable values.
+If an option is not passed, all validations pass. It accepts a value, proc, or symbol.
```
`It accepts a value, proc, or symbol` what is `It` referring here?

ChaelCodes replied:
I've updated the text to say Each option. I'm open to clearer alternatives.

ChaelCodes commented on guides/source/active_record_validations.md:
```diff
@@ -365,6 +365,36 @@ end
+### `comparison`
+
+This check will validate a comparison between any two comparable values.
+If an option is not passed, all validations pass. It accepts a value, proc, or symbol.
```
```suggestion
If an option is not passed, all validations pass. Each option accepts a value, proc, or symbol.
```

ChaelCodes commented on guides/source/active_record_validations.md:
````diff
@@ -365,6 +365,36 @@ end
+
+```ruby
+class Promotion < ApplicationRecord
+  validates :start_date, compare: { greater_than: :end_date }
````
```suggestion
  validates :start_date, comparison: { greater_than: :end_date }
```

matthewd commented on guides/source/active_record_validations.md:
```diff
@@ -365,6 +365,36 @@ end
+### `comparison`
+
+This check will validate a comparison between any two comparable values.
+If an option is not passed, all validations pass. Each option accepts a value, proc, or symbol.
```
I was going to suggest "supplied" to avoid passed/pass... but maybe it should just be an immediate error to define a comparison validation without any constraints?

ChaelCodes replied:
Do we have any existing patterns for this? I think this is the first validator that needs checks to validate anything.

I leaned towards no error in case someone dynamically supplied options, and they were missing. I also thought if there were no constraints, then there should be no errors. I can also see the argument that an error could help nudge people who've misconfigured the comparison validator.

The way I see it, we have 3 options:
:eyes: Error if no options supplied
:rocket: Pass if no options supplied
:tada: Validate if the attribute is comparable if no options supplied, making the checks optional

I don't have a strong preference, and they're all pretty easy to implement.

matthewd replied:
Sorry it's taken me a silly amount of time to come back to this... 

I believe this is what `check_validity!` is for. Numericality only does a typecheck [because it always implies the must-be-a-number constraint, even without any further options], but Length and Format have precedent for "you can supply any of these options, but must supply at least one (or for Format, exactly one).

ChaelCodes replied:
That makes sense. I'll update the tests, code, and guide.

ChaelCodes replied:
I have made the requested changes!

matthewd commented:
👋🏻!

What do you think about having NumericalityValidator inherit from this, such that it ends up being "make_sure_its_a_number && super"? 

Worth considering as a future change, or probably not worth it?

ChaelCodes commented:
@matthewd 
:wave:!

I tried out the inheritance and it ended up being more trouble than expected. With a decent refactor to NumericalityValidator though, I can add a Comparability module similar to [Clusivity.](https://github.com/rails/rails/blob/master/activemodel/lib/active_model/validations/clusivity.rb)
I'm working on that and will update soon!

ChaelCodes commented:
This PR is ready to review. Let me know if there are any other requested changes. I'd be happy to implement them.

ChaelCodes committed:
Replace all references to invalid/valid with assert_invalid_values/assert_valid_values.

ChaelCodes commented:
@rafaelfranca @matthewd I've just rebased to incorporate the new changes to NumericalityValidator, and renamed `valid!`, and `invalid!`. Please let me know if there's anything else I can do for this PR.

abhaynikam commented on activemodel/lib/active_model/validations/comparison.rb:
```diff
@@ -0,0 +1,78 @@
+      def check_validity!
+        unless (options.keys & COMPARE_CHECKS.keys).any?
+          raise ArgumentError, "Expected one of :greater_than, :greater_than_or_equal_to, "\
+          ":equal_to, :less_than, :less_than_or_equal_to, nor :other_than supplied."
```
@ChaelCodes Typo: `nor` -> `or`

I would also add `option` in the exception message like: `... or :other_than option supplied`.
